Assets/Shopify/UIToolkit/Shops/Generic/Classes/ProductListView.cs

```diff
@@ -9,17 +9,91 @@
     public class ProductListView : GenericMultiProductShopView {
         public MultiProductListItem ListItemTemplate;
         public ScrollRect ListView;
+        public GameObject content;
+
+        private const int potentiallyVisibleElements = 5;
+        private float elementWidth;
+        private int dataOffset = 0;
+
+        private bool loadingMore;
+        private bool firstLoad = true;
+
+        private List<MultiProductListItem> _scrollPoolElements = new List<MultiProductListItem> ();
+
+        public ContentRegion contentRegion;
+
+        public void Awake() {
+            contentRegion.OnScroll.AddListener (OnScroll);
+
+            RectTransform rtt = (RectTransform)ListItemTemplate.transform;
+            elementWidth = rtt.rect.width;
+
+            for (int i = 0; i < potentiallyVisibleElements; i++) {
+                var element = Instantiate (ListItemTemplate, content.transform);
+                element.gameObject.SetActive(true);
+                element.transform.localPosition = new Vector2 (elementWidth * i, 0);
+                _scrollPoolElements.Add (element);
+            }
+        }
 
         public void OnProductsLoaded(Product[] products) {
-            foreach (var product in products) {
-                var listItem = Instantiate(ListItemTemplate);
-                listItem.gameObject.SetActive(true);
-                listItem.transform.SetParent(_scrollContentRect.transform, false);
-                listItem.SetProduct(product);
-                listItem.OnClick.AddListener(() => {
-                    Shop.ViewProductDetails(product, product.variants().edges().Select((x) => x.node()).ToArray());
-                });
+            if (firstLoad) {
+                for (int i = 0; i < potentiallyVisibleElements; i++) {
+                    _scrollPoolElements[i].SetProduct (products [i]);
+                }
+
+                firstLoad = false;
             }
+
+            contentRegion.endOffset = -elementWidth * (shop.ProductCache.Count - 5);
```

Comment:
Should this 5 be the `potentiallyVisibleElements` value? If so we should use the variable instead of a constant here.